lib/ddtrace/span/duration.rb

```diff
@@ -0,0 +1,70 @@
+require 'ddtrace/utils/time'
+
+module Datadog
+  class Span
+    # Stateful class used to track and calculate the duration of a Span.
+    class Duration
+      attr_reader :start_time, :end_time
+
+      def initialize
+        @start_time = nil
+        @duration_start = nil
+        @end_time = nil
+        @duration_end = nil
+        @wall_clock_duration = false
```

Comment:
Not sure I follow why we track two different types of start/end time. I think it is because if the user supplies us a time then we don't use monotonic time?

Probably easier if we just simply track start/end time as "if provided, use that value, else use Utils::Time::get_time"?

Reply:
We don't know at monotonic start time whether the user will choose to supply their own time at finish time. if they do, then we have to fallback to wall clock. so we have to track both, is my understanding at least

Reply:
err, i changed my mind...it's messy maintaining both, as i'm already messing it up in the tests... for at least the accessors we need to abstract this away so there's only one publicly accessible start/end time

Reply:
I think you might not want to mix using monotonic and non-monotonic times. I agree it is the most-confusing part of this code, though. One of the reasons I extracted a class to manage that state! Can you mix `Utils::Time::get_time` and other provided times safely?

Reply:
yup agreed, this PR has gone thru a few iterations so some of the comments in other areas are a bit more up to date on what we're doing currently. at present afaik we don't mix monotonic and wall, we either use `duration_start - duration_end` (monotonic, although if we're running on unsupported rubies it falls back to wall) or if at any point a time is provided, we rely on  `start_time - end_time` (wall). So, at a high level, monotonic should never mix with non monotonic.